lib/memoryprofiler/src/trait.rs

```diff
@@ -0,0 +1,169 @@
+
+/// Primitive types
+impl MemoryUsage for bool {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<bool>()
+    }
+}
+impl MemoryUsage for char {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<char>()
+    }
+}
+impl MemoryUsage for f32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f32>()
+    }
+}
+impl MemoryUsage for f64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f64>()
+    }
+}
+impl MemoryUsage for i8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i8>()
+    }
+}
+impl MemoryUsage for i16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i16>()
+    }
+}
+impl MemoryUsage for i32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i32>()
+    }
+}
+impl MemoryUsage for i64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i64>()
+    }
+}
+impl MemoryUsage for isize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<isize>()
+    }
+}
+impl MemoryUsage for u8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u8>()
+    }
+}
+impl MemoryUsage for u16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u16>()
+    }
+}
+impl MemoryUsage for u32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u32>()
+    }
+}
+impl MemoryUsage for u64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u64>()
+    }
+}
+impl MemoryUsage for usize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<usize>()
+    }
+}
+
+// pointer
+// Pointers aren't necessarily safe to dereference, even if they're nonnull.
+
+// references
+impl<T: MemoryUsage> MemoryUsage for &T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+impl<T: MemoryUsage> MemoryUsage for &mut T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+
+// slices
+impl<T: MemoryUsage> MemoryUsage for [T] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// arrays
+impl<T: MemoryUsage, const N: usize> MemoryUsage for [T; N] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// strs
+impl MemoryUsage for str {
+    fn size_of(&self) -> usize {
+        self.as_bytes().size_of()
+    }
+}
+
+// TODO: tuples
+
+/// Standard library types
+
+// TODO: Arc
+
+//impl<T: MemoryUsage> MemoryUsage for Box<T> {
+//}
+
+// Cell
+
+// Is a Pin always dereferenceable?
+//impl<T: MemoryUsage> MemoryUsage for Pin<T> {
+//}
+
+// TODO: Mutex
+
+// TODO: NonNull might be possible when '*const T' is MemoryUsage.
+
+impl<T: MemoryUsage> MemoryUsage for Option<T> {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<Option<()>>() + self.iter().map(MemoryUsage::size_of).sum::<usize>()
+    }
+}
+
+// TODO: Rc
+
+// TODO: Ref, RefCell, RefMut
+
+//impl<T: MemoryUsage> MemoryUsage for Result<T> {
+//}
+
+// TODO: RwLock
+
+// string?
+
+// TODO: UnsafeCell
```

Comment:
Yeah there's actually a ton of types but we can always add them incrementally!

For example https://doc.rust-lang.org/std/num/struct.NonZeroU8.html

and the rest of std::collections

Reply:
Depending on our needs for Wasmer, too, we may also need to add features to this crate that conditionally implement this trait on third party types too, due to the orphan rule.

So for example if we use crossbeam::Mutex anywhere (which we might) then the only place this trait can be implemented is in this crate or in `crossbeam`. So the standard way to handle that is to have a feature called `crossbeam` that makes this crate depend on crossbeam and implement the trait for those types.

Reply:
I think it's clear that we should include an implementation for the `std::collections`. I initially though this crate would be separate from Wasmer, but once I thought about the orphan rule I realized that we might want it to be an implemention of the trait for all the types used in Wasmer which somewhat attaches it to wasmer.

Or we can use features as you suggest.